File: Balanced.java
import java.util.*;


public class Balanced {
    static char[][] TERMS = {{'(', ')'}, {'[', ']'}, {'{', '}'}};

    static boolean match(char openTerm, char closeTerm) {
        for (char[] c : TERMS) {
            if (c[0] == openTerm) {
                return c[1] == closeTerm;
            }
        }
        return false;
    }

    static boolean isOpenTerm(char openTerm) {
        for (char[] c : TERMS) {
            if (c[0] == openTerm) {
                return true;
            }
        }
        return false;
    }

    static boolean isBalanced(String expression) {
        Stack<Character> stack = new Stack<Character>();
        for (char c : expression.toCharArray()) {
            if (isOpenTerm(c)) {
                stack.push(c);
            }
            else {
                if (stack.isEmpty() || !match(stack.pop(), c)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    static public void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Please pass string to application");
            return;
        }
        if (isBalanced(args[0])) {
            System.out.println("Balanced");
        }
        else {
            System.out.println("Not balanced");
        }
    }
}

File: balanced.js
TERMS = ["()", "[]", "{}"];

function Stack() {
  this.data = new Array();
}

Stack.prototype.push = function (item) {
  this.data.unshift(item);
}

Stack.prototype.pop = function() {
  return this.data.shift();
}

Stack.prototype.isEmpty = function() {
  return this.data.length === 0;
}


function Balanced() {
}

Balanced.prototype.match = function(openTerm, closeTerm) {
  for (i = 0; i < 3; ++i) {
    if (TERMS[i][0] === openTerm) {
      return TERMS[i][1] === closeTerm;
    }
  }
  return false;
}

Balanced.prototype.isOpenTerm = function(openTerm) {
  for (i = 0; i < 3; ++i) {
    if (TERMS[i][0] === openTerm) {
      return true;
    }
  }
  return false;
}

Balanced.prototype.isBalanced = function(expression) {
  let stack = new Stack();
  for (i = 0; i < expression.length; ++i) {
    let c = expression[i];
    if (this.isOpenTerm(c)) {
      stack.push(c);
    }
    else {
      if (stack.isEmpty() || !this.match(stack.pop(), c)) {
        return false;
      }
    }
  }
  return stack.isEmpty();
}

test = new Balanced();
console.log(test.isBalanced('[]'));
